[1.19.2] Fixup shield synchronization

Jonathing opened the pull request:
This PR fixes up how shields are synchronized in Shield Expansion. This was done by doing the following:

- Removing the usage of the `ClearShields` packet.
  - It still exists in the mod so that clients and servers with different versions after this PR is merged will continue to work.
- Making the client clear shields when the player logs out of a world.
- Removing the local player null check in the `SyncShields` packet.
- Including an ultimate, final, fallback shield map in case the default never got synced to the client.
  - If this ever happens, a fatal log message will be sent to the console.

Additionally, a small closure was added to the buildscript to assist with IntelliJ IDEA refusing to download linked source and javadoc files.

- Fixes #47.
- Supersedes #40.
- Supersedes #48.

Jonathing committed:
Fix IntelliJ IDEA not downloading sources

Jonathing committed:
Let the client handle clearing shield data

Jonathing committed:
Handle the shield sync unconditionally

Jonathing committed:
Add an edge-case in case sync goes horribly wrong

Jonathing committed:
Made this event listener virtual because reasons

Why did you write your event listeners this way???

Jonathing committed:
Re-add sending ClearShields so we don't break older versions

SwanX1 commented on src/main/java/org/infernalstudios/shieldexp/init/ShieldDataLoader.java:
```diff
@@ -29,8 +29,10 @@
     public static Map<ResourceLocation, JsonElement> FILE_MAP = new HashMap<>();
-    public static Map<String, Map<String, Double>> SHIELD_STATS = new HashMap<>();
-    public static List<Map.Entry<ResourceLocation, JsonElement>> toSync = new ArrayList<>();
+    public static final Map<String, Map<String, Double>> SHIELD_STATS = new ShieldStatsMap();
+    public static final List<Map.Entry<ResourceLocation, JsonElement>> toSync = new ArrayList<>();
+
+    public static final String DEFAULT_SHIELD_NAME = ShieldExpansion.MOD_ID + ":default";
```
Wouldn't it be better to do this, instead of converting the RLs to Strings every time?

```suggestion
    public static final Map<ResouceLocation, Map<String, Double>> SHIELD_STATS = new ShieldStatsMap();
    public static final List<Map.Entry<ResourceLocation, JsonElement>> toSync = new ArrayList<>();

    public static final ResourceLocation DEFAULT_SHIELD_NAME = new ResourceLocation(ShieldExpansion.MOD_ID, "default");
```

Jonathing replied:
That is if you want my PR to change the semantics around `SHIELD_STATS`, which I have no problem doing. I just typically try to only add what's needed rather than changes what's existing.

SwanX1 replied:
Since your PR already changes some aspect of this, I don't see a reason to change this semantic as well.

SwanX1 commented on src/main/java/org/infernalstudios/shieldexp/init/ShieldDataLoader.java:
```diff
@@ -47,7 +49,7 @@ public void apply(Map<ResourceLocation, JsonElement> files, ResourceManager reso
 
         for (ResourceLocation name : FILE_MAP.keySet()) {
-            if (ForgeRegistries.ITEMS.containsKey(name) || name.toString().equals(ShieldExpansion.MOD_ID + ":default")) {
+            if (ForgeRegistries.ITEMS.containsKey(name) || name.toString().equals(DEFAULT_SHIELD_NAME)) {
```
See suggestion above (lines 32-35)
```suggestion
            if (ForgeRegistries.ITEMS.containsKey(name) || name.equals(DEFAULT_SHIELD_NAME)) {
```

SwanX1 commented on src/main/java/org/infernalstudios/shieldexp/init/ShieldDataLoader.java:
```diff
@@ -83,12 +85,47 @@ public static void parse(ResourceLocation name, JsonObject data) {
 
-            if (!key.equals(ShieldExpansion.MOD_ID + ":default"))
+            if (!key.equals(DEFAULT_SHIELD_NAME))
                 Config.extendList(key);
```
See suggestion above (lines 32-35)

For lines 75-89:
```diff
-       String key = name.toString();
-       if (ForgeRegistries.ITEMS.containsKey(name) || key.equals(DEFAULT_SHIELD_NAME)) {
+       if (ForgeRegistries.ITEMS.containsKey(name) || name.equals(DEFAULT_SHIELD_NAME)) {
            Map<String, Double> stats = new HashMap<>();
            stats.put("cooldownTicks", data.getAsJsonObject().get("cooldownTicks").getAsDouble());
            stats.put("speedFactor", data.getAsJsonObject().get("speedFactor").getAsDouble());
            stats.put("parryDamage", data.getAsJsonObject().get("parryDamage").getAsDouble());
            stats.put("parryTicks", data.getAsJsonObject().get("parryTicks").getAsDouble());
            stats.put("stamina", data.getAsJsonObject().get("stamina").getAsDouble());
            stats.put("blastResistance", data.getAsJsonObject().get("blastResistance").getAsDouble());
            stats.put("flatDamage", data.getAsJsonObject().get("flatDamage").getAsDouble());
-           SHIELD_STATS.remove(key);
-           SHIELD_STATS.put(key, stats);
+           SHIELD_STATS.remove(name);
+           SHIELD_STATS.put(name, stats);

-           if (!key.equals(DEFAULT_SHIELD_NAME))
-               Config.extendList(key);
+           if (!name.equals(DEFAULT_SHIELD_NAME))
+               Config.extendList(name.toString());
```

It's possible to maybe convert `Config.extendList` to take in a RL instead of a String? Not sure.

Jonathing replied:
If your config is easy to work with, I can make it happen. I am trying not to change too many semantics.

SwanX1 commented on src/main/java/org/infernalstudios/shieldexp/init/ShieldDataLoader.java:
```diff
@@ -83,12 +85,47 @@ public static void parse(ResourceLocation name, JsonObject data) {
+            // in case this map ever gets used, log a fatal error
+            @Override
+            public Double get(Object key) {
+                ShieldExpansion.LOGGER.fatal("The client is missing shield data from the server! This includes the default shield data! Please report this bug to Infernal Studios.");
```
This is not a fatal error, or an error in general, we can recover from this, it should be a warning:
```suggestion
                ShieldExpansion.LOGGER.warn("The client is missing shield data from the server! This includes the default shield data! Please report this bug to Infernal Studios.");
```

SwanX1 commented on src/main/java/org/infernalstudios/shieldexp/init/ShieldDataLoader.java:
```diff
@@ -83,12 +85,47 @@ public static void parse(ResourceLocation name, JsonObject data) {
         SHIELD_STATS.clear();
     }
+
+    private static class ShieldStatsMap extends HashMap<String, Map<String, Double>> {
```
```suggestion
    private static class ShieldStatsMap extends HashMap<ResourceLocation, Map<String, Double>> {
```

SwanX1 commented on src/main/java/org/infernalstudios/shieldexp/network/SyncShields.java:
```diff
@@ -40,9 +38,8 @@ public void handle(Supplier<NetworkEvent.Context> ctx){
-        }
+        // trust that the client can handle this even if the player isn't ready
+        // this data isn't stored to the local player so even if the player is null, it's fine
+        ShieldDataLoader.parse(shield, data.getAsJsonObject());
```
Note, not applicable to this PR: This should be reworked in the future, as using JSON to store and send these values is an unnecessary abstraction, we can just store straight doubles in the packet. The `data` object is fixed-size.

Jonathing replied:
Yes, I agree. The goal of the PR is to fixup the current system, not to reimplement it, so this solution will work for now. Honestly the correct way to do this is with some sort of custom data registry that can be synced kind of like many others in the game.

SwanX1 reviewed (changes requested):
This review may seem massive, it's just one refactor to use ResourceLocations instead of Strings in `ShieldDataLoader`. We are storing the strings as a representation of the RLs, instead of storing the RLs directly. IIRC ResourceLocations implement both `equals` and `hashCode` correctly, so we shouldn't have a problem with the comparisons and HashMaps.

Jonathing commented:
My brother, I will always appreciate a massive review. 🙏

Jonathing committed:
Calm warning message log level

Related comments:
- https://github.com/infernalstudios/Shield-Expansion/pull/49#discussion_r1885770722

Jonathing committed:
Change semantics of ShieldDataLoader.STATS_MAP

Related comments:
-
https://github.com/infernalstudios/Shield-Expansion/pull/49#discussion_r1885768724
-
https://github.com/infernalstudios/Shield-Expansion/pull/49#discussion_r1885768867
-
https://github.com/infernalstudios/Shield-Expansion/pull/49#discussion_r1885770387
-
https://github.com/infernalstudios/Shield-Expansion/pull/49#discussion_r1885770754
